### PR TITLE
feat: prefetch eth price for swap currencies

### PR DESCRIPTION
**Pull request by zzmp**
## Description

Prefetches a currency's ETH price so that USD pricing can be available instantly once a quote loads.
Cleans up the `useUSDPrice` code to make it more readable and ensure we aren't making extra calls.

_Slack thread:_ https://uniswapteam.slack.com/archives/C02T729PMQE/p1692313576887119?thread_ts=1692235420.363889&cid=C02T729PMQE

>one thing we can do is not waterfall these requests.
>
>here's an example: UNI->AAVE. In the current state:
>- enter amount (1), triggers two requests: UNI->AAVE (quote) and UNI->ETH (usd price)
>- once UNI->AAVE loads, AAVE->ETH will load (usd price)
>in the proposed state, https://github.com/Uniswap/interface/pull/7187:
>- on token selection, triggers request: TOKEN->ETH; in this case, UNI->ETH (usd price) and AAVE->ETH (usd price)
>- enter amount (1), triggers one request: UNI->AAVE (quote)
>- once entered, and once quoted, pricing is prefetched and renders instantly
>
>demo links:
>- current: https://app.uniswap.org/#/swap?inputCurrency=0x1f9840a85d5af5bf1d1762f925bdaddc4201f984&outputCurrency=0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9
>- proposed: https://interface-git-audit-pricing-uniswap.vercel.app/swap?inputCurrency=0x1f9840a85d5af5bf1d1762f925bdaddc4201f984&outputCurrency=0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9

### Before
|    Before    |  After    |
| ------------ | ------------ |
| <video src="https://github.com/Uniswap/interface/assets/5403956/8e7d44ac-3fb3-409e-990e-1b1c5f5a0c75"> | <video src="https://github.com/Uniswap/interface/assets/5403956/89dc34fa-6d55-4106-a11c-d1fad456fa1f"> |

### Automated testing

:( there are no tests for these files, and we do not have a way (yet) to test for performance regressions :(

### CODE REVIEW

**Review by JackShort — APPROVED** [review]

File: src/pages/Swap/index.tsx [hunk 1]
```diff
@@ -299,8 +299,8 @@ export function Swap({
     [independentField, parsedAmount, showWrap, trade]
   )
 
-  const fiatValueInput = useUSDPrice(parsedAmounts[Field.INPUT])
-  const fiatValueOutput = useUSDPrice(parsedAmounts[Field.OUTPUT])
+  const fiatValueInput = useUSDPrice(parsedAmounts[Field.INPUT], currencies[Field.INPUT] ?? undefined)
```
Comment (cbachmeier) [inline src/pages/Swap/index.tsx:302]: Can we also add this improvement to the swap modal in the Token Details page to maintain parity? see https://app.uniswap.org/#/tokens/ethereum/0x1f9840a85d5af5bf1d1762f925bdaddc4201f984

File: src/hooks/useUSDPrice.ts [hunk 2]
```diff
@@ -19,66 +20,82 @@ const ETH_AMOUNT_OUT: { [chainId: number]: CurrencyAmount<Currency> } = {
...
-
-  return { data: parseFloat(ethValue.toExact()) * ethUSDPrice, isLoading: false }
+  return useMemo(() => {
+    if (!currencyAmount) {
+      return { data: undefined, isLoading: false }
+    } else if (stablecoinPrice) {
+      return { data: parseFloat(stablecoinPrice.quote(currencyAmount).toSignificant()), isLoading: false }
+    } else {
+      // Otherwise, get the price of the token in ETH, and then multiple by the price of ETH
+      const ethUSDPrice = data?.token?.project?.markets?.[0]?.price?.value
```
[the remark this answers is not in the document]
Comment (zzmp) [reply src/hooks/useUSDPrice.ts:93]: no, this is eth's usd price. the token's value in eth is ethValue.

[the remark this answers is not in the document]
Comment (zzmp) [reply src/hooks/useUSDPrice.ts:93]: updating vars for clarity

File: src/pages/Swap/index.tsx (hunk 1 above)
Comment (zzmp) [reply src/pages/Swap/index.tsx:302]: I think they already share code

Comment (cbachmeier) [reply src/pages/Swap/index.tsx:302]: they don't you still have the sequential loading state on details for usd price

**Review by cbachmeier — APPROVED** [review]

**Comment by natew** [comment]
Very cool, makes me think we could do a variety of pre fetching, like when they hover over a currency kick of some sort of likely expected swap fetches or something, not sure how expensive our quote calls are.

**Pull request merged**